### PR TITLE
Export solutions

### PR DESCRIPTION
**Pull request by ErikSchierboom**
This PR restores the functionality from v2 to export the files of the last 500 solutions, which can be very helpful with representers or analyzers.

CC @junedev

### CODE REVIEW

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [inline app/commands/exercise/export_solutions_to_zip_file.rb]: The v2 version of this code flattened all filenames (effectively removing any sub-directories). I've fixed this in this version.

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [inline app/commands/exercise/export_solutions_to_zip_file.rb]: The v2 version of this code wrote files to the file system to then zip the directory recursively. With this version, we'll directly write to the zip file without the use of any intermediate files or recursive directory zipping helper needed.

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [inline app/commands/exercise/export_solutions_to_zip_file.rb:35]: I've omitted solutions that have the `:started` status.

File: app/controllers/api/solutions/export_controller.rb
Comment (ErikSchierboom) [inline app/controllers/api/solutions/export_controller.rb]: I've only made this functionality available for maintainers and admins. 

So far, I've not yet created a UI in the `/maintaining` part of the website.

File: config/routes/api.rb
Comment (ErikSchierboom) [inline config/routes/api.rb]: I've considered various paths, but felt that `/api/v2/tracks/csharp/exercises/leap/export_solutions` was nice. Let me know what you think.

File: config/initializers/rack_attack.rb
Comment (ErikSchierboom) [inline config/initializers/rack_attack.rb]: Now that I had two throttle blocks, I decided that it might be nice to extract this shared logic and monkey patch it.

File: config/initializers/rack_attack.rb
Comment (ErikSchierboom) [inline config/initializers/rack_attack.rb]: I've changed this to do route name matching, which is less likely to be wrong than a regex.

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [inline app/commands/exercise/export_solutions_to_zip_file.rb]: I've used the v2 version of the code as inspiration, but I was wondering whether we might want to output the files like the representer/analyzer would receive them (which would include test files and such).

File: config/routes/api.rb
Comment (iHiD) [reply config/routes/api.rb]: Yeah, I'm good with this. But I'd make the controller match it (ie ExportSolutionsController).

File: app/controllers/api/solutions/export_controller.rb
Comment (iHiD) [inline app/controllers/api/solutions/export_controller.rb]: I'd have a slight preference here to having a `use_track` method which is:

```ruby
def use_track
  @track = Track.find(params[:track_slug])
rescue ActiveRecord::RecordNotFound
  render_track_not_found
end
```

and then the same for exercise

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (iHiD) [inline app/commands/exercise/export_solutions_to_zip_file.rb]: `solution.latest_iteration `

This should be the first iteration, not the latest one (we want to see what they look like pre-mentoring)

Comment (iHiD) [reply app/commands/exercise/export_solutions_to_zip_file.rb]: With a maintainer hat on, I'd probably rather just get the students' files as those are the ones that I have to deal with. But if your maintainers hat (or other people) disagree, I don't have any strong preference. Do we actually pass test files etc to representers?

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (iHiD) [inline app/commands/exercise/export_solutions_to_zip_file.rb:28]: I always find it so weird that we have to manually do this!

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [reply app/commands/exercise/export_solutions_to_zip_file.rb]: > Do we actually pass test files etc to representers?

We do. All exercise files are passed to the tooling, except for the files within `.docs` (and the solution files overwrite any exercise files).

> But if your maintainers hat (or other people) disagree,

The reason why we do the above, is that some tracks might have non-solution files (e.g. project files) that are needed for the representer/analyzer to work well. E.g. the F# analyzer might want to analyze the project file for the order of file dependencies, which is not a solution file.

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [reply app/commands/exercise/export_solutions_to_zip_file.rb:28]: Yes!

File: config/routes/api.rb
Comment (ErikSchierboom) [inline config/routes/api.rb]: @iHiD Is this the way you'd like the controller to be named?

File: app/commands/exercise/export_solutions_to_zip_file.rb
Comment (ErikSchierboom) [inline app/commands/exercise/export_solutions_to_zip_file.rb]:
```suggestion
        # Export the other (exercise) files
```

File: config/routes/api.rb
Comment (iHiD) [reply config/routes/api.rb]: Yes, which means you can leave the `controller` param out altogether on this line.

File: app/controllers/api/export_solutions_controller.rb
Comment (iHiD) [inline app/controllers/api/export_solutions_controller.rb]: You didn't like my suggestion of `find` and the rescue? (https://github.com/exercism/website/pull/2847#discussion_r943327471)

Comment (ErikSchierboom) [reply app/controllers/api/export_solutions_controller.rb]: I missed the rescue part. I'll update

File: app/controllers/api/export_solutions_controller.rb
Comment (iHiD) [inline app/controllers/api/export_solutions_controller.rb]:
```suggestion
      @track = Track.find(params[:track_slug])
```

Comment (iHiD) [reply app/controllers/api/export_solutions_controller.rb]: The `FriendlyId` gem means that we can just use `find` here.

File: app/controllers/api/export_solutions_controller.rb
Comment (iHiD) [inline app/controllers/api/export_solutions_controller.rb]:
```suggestion
      @exercise = @track.exercises.find(params[:exercise_slug])
```

**Review by iHiD — APPROVED** [review]
I'll let you merge this on Tues in case rack attack causes issues.

**Pull request merged**
